### PR TITLE
Add use_tutorial() function, documentation, tests

### PR DESCRIPTION
**Pull request by angela-li**
As mentioned in #314, with big shoutouts to @xvrdm @jules32 @jessesadler

We love tidyverse dev day!!

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -46,7 +46,8 @@ Suggests:
     roxygen2,
     spelling (>= 1.2),
     styler (>= 1.0.2),
-    testthat (>= 2.0.0)
+    testthat (>= 2.0.0),
+    learnr
```
Comment (hadley) [inline DESCRIPTION]: I don't think usethis needs learnr?

File: R/tutorial.R [hunk 2]
```diff
@@ -0,0 +1,44 @@
+#' Create a tutorial.
+#'
+#' Creates new tutorial in `inst/tutorials`. Tutorials are interactive R Markdown documents built with the `learnr` package.
```
Comment (hadley) [inline R/tutorial.R]: Can you please make sure this line is wrapped to ~80 characters? (I think it's a bit long currently)

File: tests/testthat/test-use-tutorial.R [hunk 3]
```diff
@@ -0,0 +1,39 @@
+context("test-use-tutorial")
+
+test_that("use_tutorial() requires a `name`", {
```
Comment (hadley) [inline tests/testthat/test-use-tutorial.R]: I think you can combine this with the next test into something like "use_tutorial() checks its inputs"

File: tests/testthat/test-use-tutorial.R [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+})
+
+test_that("use_tutorial() requires a `title`", {
+  skip_if_not_installed("rmarkdown")
+
+  scoped_temporary_package()
+  expect_error(use_tutorial(name = "tutorial-file"), "no default")
+})
+
+test_that("use_tutorial() creates a tutorials folder", {
```
Comment (hadley) [inline tests/testthat/test-use-tutorial.R]: I think you could also combine this with the next two tests

File: R/tutorial.R [hunk 5]
```diff
@@ -0,0 +1,44 @@
+#' Create a tutorial.
```
Comment (jennybc) [inline R/tutorial.R]:
```suggestion
#' Create a learnr tutorial
```

File: R/tutorial.R (hunk 2 above)
Comment (jennybc) [inline R/tutorial.R]:
```suggestion
#' Creates a new tutorial in `inst/tutorials`. Tutorials are interactive R
#' Markdown documents built with the learnr package.
```

File: R/tutorial.R [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+#'
+#' Creates new tutorial in `inst/tutorials`. Tutorials are interactive R Markdown documents built with the `learnr` package.
+#'
+#' @section General setup:
+#' * Adds needed packages to `DESCRIPTION`.
+#' * Adds `inst/tutorials/*.html` to `.gitignore` so
+#'   you never accidental track rendered tutorials.
```
Comment (jennybc) [inline R/tutorial.R]:
```suggestion
#'   you never accidentally track rendered tutorials.
```

File: R/tutorial.R [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+#' @param name Base for file name to use for new tutorials. Should consist only
+#'   of numbers, letters, _ and -. We recommend using lower case.
+#' @param title The title of the tutorial
+#' @inheritParams use_template
+#' @seealso The [`learnr` package documentation](https://rstudio.github.io/learnr/index.html).
```
Comment (jennybc) [inline R/tutorial.R]:
```suggestion
#' @seealso The [learnr package documentation](https://rstudio.github.io/learnr/index.html).
```

File: R/tutorial.R [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+#' use_tutorial("learn-to-do-stuff", "Learn to do stuff")
+#' }
+use_tutorial <- function(name, title, open = interactive()) {
+  stopifnot(is_string(name))
+  stopifnot(is_string(title))
+
+  dir_path <- path("inst", "tutorials")
+
+  use_directory(dir_path)
+  use_git_ignore(c("*.html"), directory = dir_path)
```
Comment (jennybc) [inline R/tutorial.R]:
```suggestion
  use_git_ignore("*.html", directory = dir_path)
```

File: R/tutorial.R [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  new <- use_template(
+    "tutorial-template.Rmd",
+    save_as = path,
+    data = data,
+    ignore = TRUE,
+    open = open
+  )
+  if (!new) return(invisible(FALSE))
+
+  invisible(TRUE)
```
Comment (jennybc) [inline R/tutorial.R]: I think these lines can be simplified to `invisible(new)`.

**Comment by jules32** [comment]
Hi @hadley and @jennybc! Thanks for the comments and suggestions. We haven't had the chance to work on this yet, but we haven't forgotten! We'll work on it as soon as we can!

**Comment by xvrdm** [comment]
Happy to help out if you want to check this together.

**Comment by garrettgman** [comment]
@angela-li @jules32 @xvrdm @jessesadler This is awesome work—I really like this feature! In fact, I think I could begin using it immediately once you finish.

**Comment by jules32** [comment]
Hi everyone! 

@allisonhorst and I have spent some time going through your comments (thank you!) and trying to understand GitHub's review process, which we are encountering for the first time. It looks like we need write privileges to @angela-li's repo to make the suggested changes (otherwise we'll have to fork Angela's repo and create a pull request which seems like an additional layer of complexity?) If that is correct, @angela-li could you give us all permission, and we could try to have a go? Thanks!

**Comment by jennybc** [comment]
Yeah, to make changes to this PR, one needs write (push) permission to her fork.

You'll have to then keep working in her `master` branch, which is regrettably where this PR comes from.

In the future, it's much better for everyone to make PRs from a non-`master` branch.

https://happygitwithr.com/fork-and-clone.html#dont-touch-master

**Comment by jennybc** [comment]
If you end up needing (or wanting) to pull changes from `master` of this repo (the main repo) into your PR, it is conceivable that you should make a fresh PR from a non-`master` branch of somebody's fork:

  1. Pick someone's fork as your workspace. Give the other people on the PR team write (push) access. Let us call this the Chosen Fork.
  1. Not strictly necessary, but easiest: Make sure Chosen Fork's `master` branch is synced with `master` here. True by definition for a new fork but might require [pull upstream changes](https://happygitwithr.com/upstream-changes.html) for a pre-existing fork.
  1. Everyone on the PR team adds The Chosen Fork as a remote.
  1. Someone creates a new branch for the new PR with basis = current `HEAD` of `master` here in `tidyverse/usethis`, ports the work from this PR over to that new branch, and pushes the branch to the Chosen Fork.
  1. Now the PR team can work collaboratorively in that branch.
  1. When you're ready, make a new PR and link to this conversation.

Hope that helps. Some of this complication comes from making the initial PR from `master`. But some is unavoidable when a collaborative, external team is working on a PR.

**Comment by angela-li** [comment]
Hey folks! First, apologies for the PR from master. I realized my mistake
at tidyverse dev day and was going to revise it to make it from a branch. I
asked Mine at tidyverse dev day about it — I think she may have
misunderstood my question and told me to go ahead and submit it the way it
was (oops)!

Second, apologies for letting this PR languish so long! I’ll take some time
today, clean it up and submit from a new synced-with-master branch of my
fork. I’ll also add everyone on the PR as collaborators to my fork so you
all can contribute in the future.

Thanks all for letting us have this learning experience and for the awesome
Git guidance, I appreciate it!

**Comment by jules32** [comment]
Thanks so much @jennybc for the explanation, it is super helpful and I appreciate you writing it out. 

And thanks @angela-li for taking the next steps with it; let us know how to help!

**Comment by angela-li** [comment]
Closing as PR #645 incorporates suggested edits, and is properly from a branch! Thanks all for your patience.

**Comment by jules32** [comment]
Awesome, thanks @angela-li! It was really great to work with everyone here on this, thanks!

**Pull request closed without merging**